Validate JWT tokens in the ExecutionAPI for strong task identity

ashb opened the pull request:
This adds in Bearer auth token validation for the entire Execution API server,
and JWT generation to the Scheduler before handing off to the executor.

Some notes about how this is achieved:

* We make heavy use of FastAPI's Dependency Injection to make this work

* We use the [`svcs`][1] module to make managing the service, and viewing the
  state of it, easy

* In `.execution_api.routes` there are now two routers used -- one which
  ensures that _everything_ underneath it has a valid ExecutionAPI JWT, and
  another one which is "public" (and should contain very few routes)

* We ran into an issue (linked in code) with `__future__ annotations` imports,
  and had to disable it on one specific file.

* Refreshing of expiring Task Identity tokens is not handled yet (this will
  need middleware to support, a Dependency cannot add the response headers)

* This could almost certainly use more tests than it has right now.

* The API docs know that a JWT Bearer is needed
  <img width="780" alt="Image" src="https://github.com/user-attachments/assets/2568a7e4-1812-42b8-a018-c3b1618111ad" />

Closes #45107

[1]: pypi.org/project/svcs

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

ashb committed:
Validate JWT tokens in the ExecutionAPI for strong task identity

This adds in Bearer auth token validation for the entire Execution API server.

Some notes about how this is achieved:

* We make heavy use of FastAPI's Dependency Injection to make this work

* We use the `svcs`[1] module to make managing the service, and viewing the
  state of it, easy

* In `.execution_api.routes` there are now two routers used -- one which
  ensures that _everything_ underneath it has a valid ExecutionAPI JWT, and
  another one which is "public" (and should contain very few routes)

* We ran into an issue (linked in code) with `__future__ annotations` imports,
  and had to disable it on one specific file.

* Refreshing of expiring Task Identity tokens is not handled yet (this will
  need middleware to support, a Dependency cannot add the response headers)

[1]: pypi.org/project/svcs

kaxil commented on airflow/api_fastapi/execution_api/routes/health.py:
```diff
@@ -17,11 +17,33 @@
+    failing: dict[str, str] = {}
+    code = 200
+
+    print(services)
```
```suggestion
```

kaxil commented on airflow/api_fastapi/execution_api/routes/xcoms.py:
```diff
@@ -20,26 +20,48 @@
+    """Check if the task has access to the XCom."""
+    # TODO: Placeholder for actual implementation
+
+    write = request.method not in {"GET", "HEAD", "OPTIONS"}
```
Should we include DELETE too?

ashb replied:
DELETE counts as a write. There are more write/mutating verbs that read-only ones. `not in` to the rescue

kaxil replied:
oh damn, I read it completely wrong.

ashb commented on airflow/executors/base_executor.py:
```diff
@@ -128,8 +130,32 @@ class BaseExecutor(LoggingMixin):
+        from airflow.configuration import conf
+
+        generator = JWTGenerator(
+            valid_for=conf.getint("jwt_expiration_time", "jwt_expiration_time"),
```
```suggestion
            valid_for=conf.getint("execution_api", "jwt_expiration_time"),
```

pierrejeambrun commented on tests/api_fastapi/execution_api/conftest.py:
```diff
@@ -16,12 +16,23 @@
+        auth = AsyncMock(spec=JWTValidator)
+        auth.avalidated_claims.return_value = {"sub": "edb09971-4e0e-4221-ad3f-800852d38085"}
+
+        # Inject our fake JWTValidaotr object. Can be over-ridden by tests if they want
```
```suggestion
        # Inject our fake JWTValidator object. Can be over-ridden by tests if they want
```

pierrejeambrun commented on airflow/api_fastapi/execution_api/app.py:
```diff
@@ -34,10 +37,33 @@
+    # record this here
+    app.state.svcs_registry = registry
+
+    # Create an app scoped validator, so that we don't have to fetch it every time
+    registry.register_value(JWTValidator, _jwt_validator(), ping=JWTValidator.status)
```
I feel like introducing `svcs` just for caching `JWTValidator` might be a little bit overkill. (especially when we can achieve the same with fastapi and dependecy_override for tests) but I might be missing something.

(also not strongly against but I feel like this should be used if the usage was a little bit more intensive and spread outside of fastapi only)

ashb replied:
This is a low-impact way of introducing svcs to see what it feels like and if it's worth using larger (for instance using it for DB session management)

pierrejeambrun reviewed (approved):
Nice I think it's a solid base to keep iterating on, multiple things could be improved but I guess this will be done in follow up PR.

just one remark but not blocking

ashb committed:
Set `api_auth/jwt_secret` at deploy-time in our Kube tests

Without this being set the API server now crash-loops on startup, which adds
an extra 2 minutes to the test time for no reason.

ashb committed:
Update tests/api_fastapi/execution_api/conftest.py

Co-authored-by: Pierre Jeambrun <pierrejbrun@gmail.com>
